add blog post for release of hpy-0.0.4 (#24)
hpyproject/hpyproject.org: posts/2021/10/hpy-0.0.4.md

<!--
.. title: hpy 0.0.4: Third public release
.. slug: hpy-0.0.4-third-public-release
.. date: 2021-10-22 15:00:00 UTC
.. author: fangerer
.. tags:
.. category:
.. link:
.. description:
.. type: text
-->

HPy 0.0.4 is out! The third official HPy release comes with many new features
and was again made available on PyPI (first time with binary wheels).

Besides new API functions, the major highlights of this release are Python 3.10
support, better support for native fields (HPyField) and global variables

> (HPyGlobal), 

Reviewer: I would add a few more things to this list -- maybe mention one or two of the new API functions as examples and that there are new features in debug mode. Otherwise the introduction makes it sound like not too much happened when a lot did happen.

> Great new also is that we are now able to provide two more non-trivial projects
> that have been (partially) migrated to HPy. This is,
> [Kiwisolver](https://github.com/hpyproject/kiwi-hpy/)
> and [Matplotlib](https://github.com/hpyproject/matplotlib-hpy/).

Reviewer: ```suggestion
Great news too is that we are now able to provide two more non-trivial projects
that have been (partially) migrated to HPy. This is,
[Kiwisolver](https://github.com/hpyproject/kiwi-hpy/)
and [Matplotlib](https://github.com/hpyproject/matplotlib-hpy/).
```

Reviewer 2: I would also add `We are still cleaning these ports up and will write another blog post about the ports and open them for discussion with the project owners.`

which is used to autogenerate parts of the HPy code and is a reliable list of
all the supported functions. Also have a look at additional helpers in
[`inline_helpers.h`](https://github.com/hpyproject/hpy/blob/release/0.0.4/hpy/devel/include/hpy/inline_helpers.h).

!!! Warning

    The HPy API is still considered in alpha status and it's subject to change
    between versions.

Debug Mode
==========

We again improved HPy's debug mode and added following new features:

Detect Invalid Use of Raw Data Pointers
---------------------------------------

Some API functions return a raw data pointer from an object. For example:
```
const char* HPyUnicode_AsUTF8AndSize(HPyContext *ctx, HPy h, HPy_ssize_t *size)
```
returns a raw data pointer to the UTF8 representation of a Python unicode
object. HPy doesn't expose the internal representation of the unicode object, so
the Python implementation may use an arbitrary internal representation. This
means that the UTF8 representation is just temporarily created for this API call
and so the raw data must be released at some point. The contract here is that

> the raw data pointer is as long valid as the corresponding handle is valid.
> Example:

Reviewer: ```suggestion
the raw data pointer is valid as long as the corresponding handle is valid.

Example:
```

```c
#include <string.h>

static const char *s_hello_world = "Hello, World!";

static const char * foo(HPyContext *ctx)
{
    HPy h_unicode = HPyUnicode_FromString(ctx, s_hello_word);
    HPy_ssize_t size;
    const char *res = HPyUnicode_AsUTF8AndSize(ctx, h_unicode, &size);

    /* closing 'h_unicode' is, of course, correct */
    HPy_Close(ctx, h_unicode);

    /* raw data pointer 'res' may have become invalid when closing
       'h_unicode' */
    return res;
}

static int bar(HPyContext *ctx)
{
    const char *s = foo(ctx);

    /* accessing 's' will cause a fatal error in debug mode (on supported
    systems) */
    return strcmp(s, s_hello_world) == 0;
}
```

> It
> is easy to forget about this resriction and if the raw data pointer is used

Reviewer: ```suggestion
It is easy to forget about this resriction and if the raw data pointer is used
```
